src/bls12_381/fp2.rs

```diff
@@ -0,0 +1,1236 @@
+                15116930867080254631,
+                3288288975085550621,
+                1021049300055853010,
+            ]),
+        };
+        // 7. c7 = Z^((c2 + 1) / 2)
+        const C7: Fp2 = Fp2 {
+            c0: Fp::from_raw_unchecked([
+                1921729236329761493,
+                9193968980645934504,
+                9862280504246317678,
+                6861748847800817560,
+                10375788487011937166,
+                4460107375738415,
+            ]),
+            c1: Fp::from_raw_unchecked([
+                16821121318233475459,
+                10183025025229892778,
+                1779012082459463630,
+                3442292649700377418,
+                1061500799026501234,
+                1352426537312017168,
+            ]),
+        };
+        let mut tv1 = C6; // 1. tv1 = c6
+        let tv2 = div.pow_vartime(C4); // 2. tv2 = v^c4
+        let tv3 = tv2.square(); // 3. tv3 = tv2^2
+        let tv3 = tv3 * div; // 4. tv3 = tv3 * v
+        let tv5 = num * tv3; // 5. tv5 = u * tv3
+        let tv5 = tv5.pow_vartime(C3); // 6. tv5 = tv5^c3
+        let tv5 = tv5 * tv2; // 7. tv5 = tv5 * tv2
+        let tv2 = tv5 * div; // 8. tv2 = tv5 * v
+        let mut tv3 = tv5 * num; // 9. tv3 = tv5 * u
+        let mut tv4 = tv3 * tv2; // 10. tv4 = tv3 * tv2
+        let tv5 = tv4.pow_vartime(C5); // 11. tv5 = tv4^c5
+        let is_square = tv5.ct_eq(&Fp2::one()); // 12. isQR = tv5 == 1
+        let tv2 = tv3 * C7; // 13. tv2 = tv3 * c7
+        let tv5 = tv4 * tv1; // 14. tv5 = tv4 * tv1
+        tv3.conditional_assign(&tv2, !is_square); // 15. tv3 = CMOV(tv2, tv3, isQR)
+        tv4.conditional_assign(&tv5, !is_square); // 16. tv4 = CMOV(tv5, tv4, isQR)
+                                                  // 17. for i in (c1, c1 - 1, ..., 2):
+        for i in (2..=C1).rev() {
+            let tv5 = i as u32 - 2; // 18.    tv5 = i - 2
+            let tv5 = num_bigint::BigUint::from(2u64).pow(tv5); // 19.    tv5 = 2^tv5
+            let tv5 = tv4.pow_vartime(&tv5.to_u64_digits()); // 20.    tv5 = tv4^tv5
+            let e1 = tv5.ct_eq(&Fp2::one()); // 21.    e1 = tv5 == 1
+            let tv2 = tv3 * tv1; // 22.    tv2 = tv3 * tv1
+            tv1 = tv1 * tv1; // 23.    tv1 = tv1 * tv1
+            let tv5 = tv4 * tv1; // 24.    tv5 = tv4 * tv1
+            tv3.conditional_assign(&tv2, !e1); // 25.    tv3 = CMOV(tv2, tv3, e1)
+            tv4.conditional_assign(&tv5, !e1); // 26.    tv4 = CMOV(tv5, tv4, e1)
+        }
+        (is_square, tv3)
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Fp2Bytes {
+    pub slice: [u8; 96],
+}
+
+impl Default for Fp2Bytes {
+    fn default() -> Self {
+        Self { slice: [0u8; 96] }
+    }
+}
+
+impl AsMut<[u8]> for Fp2Bytes {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.slice
+    }
+}
+
+impl AsRef<[u8]> for Fp2Bytes {
+    fn as_ref(&self) -> &[u8] {
+        &self.slice
+    }
+}
+
+impl ff::PrimeField for Fp2 {
```

Comment:
Is this needed somewhere? It is weird to implement `PrimeField` for `Fp2` because it is not a prime field. 

I don't think this should be needed for `halo2-ecc` usage.

Reply:
Referring to the previous comment, since `Fp2` must implement `WithSmallOrderMulGroup` to be used as `G2Affine` base it's also must implement `PrimeField` https://github.com/zkcrypto/ff/blob/main/src/lib.rs#L358

Reply:
I agree that it's might become confusion, so perhaps I can add commends explaining the technical reasons why `PrimeField` and  `WithSmallOrderMulGroup` traits needed to be implemented for Fp2